manufacturer-toolkit/src/main/java/org/sdo/sct/mt/OnDieCache.java

```diff
@@ -0,0 +1,239 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Attribute;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+
+
+@Service
+public class OnDieCache {
+
+  private boolean autoUpdate = false;
+
+  private String cacheDir = "";
+
+  private final List<URL> sourceUrl = new ArrayList<URL>();
+
+  private HashMap<String, byte[]> cacheMap = new HashMap<String, byte[]>();
+
+  private final String cacheUpdatedTouchFile = "cache_updated";
+
+  /**
+   * Constructor.
+   *
+   * @param cacheDir cacheDir
+   * @param autoUpdate autoUpdate
+   * @param sourceUrlList sourceUrlList
+   */
+  @Autowired
+  public OnDieCache(@Value("${sdo.ondiecache.cachedir:null}") final String cacheDir,
+                    @Value("${sdo.ondiecache.autoupdate:false}") final boolean autoUpdate,
+                    @Value("${sdo.ondiecache.urlsources:}") final String sourceUrlList)
+      throws Exception {
+
+    if (sourceUrlList != null && !sourceUrlList.isEmpty()) {
+      String[] urls = sourceUrlList.split(",");
+      for (String url : urls) {
+        this.sourceUrl.add(new URL(url.trim()));
+      }
+    } else {
+      // defaults: the public facing sites containing OnDie artifacts
+      this.sourceUrl.add(new URL("https://tsci.intel.com/content/OnDieCA/certs/"));
+      this.sourceUrl.add(new URL("https://tsci.intel.com/content/OnDieCA/crls/"));
+    }
+
+    if (cacheDir != null) {
+      File cache = new File(cacheDir);
+      if (!cache.exists()) {
+        throw new IOException("OnDieCertCache: cache directory does not exist: " + cacheDir);
+      }
+      if (!cache.isDirectory()) {
+        throw new IOException("OnDieCertCache: cache directory must be a directory: " + cacheDir);
+      }
+
+      this.cacheDir = cacheDir;
+      this.autoUpdate = autoUpdate;
+      if (autoUpdate) {
+        // update local cache
+        copyFromUrlSources();
+      }
+      loadCacheMap();
+    }
+  }
+
+
+  /**
+   * Copy the certs and CRLs from the URL sources to the cache directory.
+   *
+   */
+  private void copyFromUrlSources() {
+
+    File cache = new File(cacheDir);
+    if (!cache.exists()) {
+      cache.mkdir();
+    }
+
+    try {
+      for (URL url : this.sourceUrl) {
+        URLConnection urlConn = url.openConnection();
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(
+            urlConn.getInputStream()))) {
+
+          // loop through all the href entries and for each .crl and .cer
+          // links, download the file and store locally
+          Document doc = Jsoup.connect(url.toString()).get();
```

Comment:
Considering we are going to download a considerable chunk of CRL/CERT files, we should document it in storage requirement.

Reply:
it turns out that there is only 40K of data in 76 files. Unlike EPID the number of certificates is very few. Nevertheless, I agree we should document it.